### PR TITLE
Integrated 1D BRD

### PR DESCRIPTION
**Pull request by aabeaton**

### CODE REVIEW

**Comment by aabeaton** [comment]
Ok, I have gone through each commit and am cool with everything here

File: pyspecdata/core.py [hunk 1]
```diff
@@ -4822,7 +4832,8 @@ def mod_BRD(guess,maxiter=20):
             logger.debug(strm('K dimlabels',K.dimlabels,'and raw shape',K.data.shape))
             self.reorder(dimname, first=False) # make the dimension we will be regularizing innermost
             logger.debug(strm('shape of the data is',ndshape(self),'is it fortran ordered?'))
-            data_fornnls = self.data.real
+            data_fornnls = self.data
+            demand_real(data_fornnls)
```
Comment (jmfranck) [inline pyspecdata/core.py]: This is more of a strategy/philosophy thing here --> I don't want to potentially "silently" convert the data to real -- instead, I try to raise a really explicit error message letting the user know what's going on, so they can do it themselves.

File: pyspecdata/core.py [hunk 2]
```diff
@@ -4591,6 +4591,16 @@ def nnls(self, dimname, newaxis_dict, kernel_func, l=0):
...
+        def demand_real(x, addtxt=''):
+            if not x.dtype == real64:
+                if x.dtype == complex128:
+                    raise ValueError("you are not allows to pass nnls complex data:\nif it makes sense for you, try yourdata.real.nnls( where you now have yourdata.nnls("+'\n'+addtxt)
+                else:
+                    raise ValueError("I expect double-precision floating point (float64), but you passed me data of dtype "+str(x.dtype)+'\n'+addtxt)
+        demand_real(self.data)
+        demand_real(self.getaxis(dimname),"(this message pertains to the %s axis)"%dimname)
+        for k,v in newaxis_dict.items():
+            demand_real(v,"(this message pertains to the new %s axis)"%str(k))
```
Comment (jmfranck) [inline pyspecdata/core.py:4603]: following on your comments, I'm trying to check that _all_ the important things are real

File: demos/nddata_NNLS_Tikhonov_190104.py [hunk 3]
```diff
@@ -18,268 +18,268 @@
...
+#
+## Final result for 1.5D test data
+#
+#fl.next(r'show result where $\lambda$ set to knee')
+#result = test_data_2d.C.nnls('t_indirect',R,
+#                    lambda x,y: exp(-y*x), l=heel_lambda)
+#fl.image(result)
+#
+#
+## ## Testing 2D BRD extension
```
Comment (jmfranck) [inline demos/nddata_NNLS_Tikhonov_190104.py:282]: please don't check in commented code

I do think it's useful to have a test, but can we make this a script (in case this is a notebook) and put in the "examples" folder -- I would like to get rid of the demo folder

File: examples/BRD_test.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+from pyspecdata import *
+from scipy.optimize import nnls
+init_logging('debug')
+vd_list = nddata(linspace(5e-4,10,25),'vd')
+t1_name = r'$\log(T_1)$'
+logT1 = nddata(r_[-4:2:100j],t1_name)
+def Gaussian_1d(axis,mu1,sigma1):
+    this_G = exp(-(axis-mu1)**2/2/sigma1**2)
+    return this_G
+true_F = Gaussian_1d(logT1.C.run(lambda x: 10**(x)),6,0.3)
```
Comment (jmfranck) [inline examples/BRD_test.py:12]: OK -- so actually this looks like an example -- does the file in "demos" do anything this does not?

File: examples/matrix_mult.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
+# -*- coding: utf-8 -*-
+from pyspecdata import *
+#from scipy.signal import gaussian
+from scipy.special import jv
+from scipy.interpolate import UnivariateSpline
+from scipy.signal import tukey
```
Comment (jmfranck) [inline examples/matrix_mult.py:6]: @bett8500 : these are not used in this example, right?

File: pyspecdata/core.py [hunk 6]
```diff
@@ -2593,8 +2594,10 @@ def thiserrbarplot(*tebargs,**tebkwargs):
     return retval
 #}}}
 #{{{general functions
-def box_muller(length, return_complex=True):
+def box_muller(length, return_complex=True, seed=None):
     r'''algorithm to generate normally distributed noise'''
+    if seed is not None:
+        pl_seed(seed)
```
Comment (jmfranck) [inline pyspecdata/core.py:2600]: you shouldn't have to do it this way -- if you seed, the seed should be global (e.g. you can seed in your script), I think

File: pyspecdata/core.py [hunk 7]
```diff
@@ -3521,14 +3525,30 @@ def get_plot_color(self):
...
+        :func:`~pyspecdata.nddata.along` to specify the dimensions for matrix
+        multiplication / dot product.
 
         >>> a = nddata(r_[0:9],[3,3],['a','b'])
         >>> b = nddata(r_[0:3],'b')
         >>> print a.C.dot(b)
         >>> print a.data.dot(b.data)
-
+)
         >>> a = nddata(r_[0:27],[3,3,3],['a','b','c'])
```
Comment (jmfranck) [inline pyspecdata/core.py:3552]: what's this paren?

File: pyspecdata/core.py [hunk 8]
```diff
@@ -3538,32 +3558,67 @@ def dot(self,arg):
         >>> b = nddata(r_[0:9],[3,3],['a','d'])
         >>> print a.C.dot(b)
         >>> print tensordot(a.data,b.data,axes=((0),(0)))
+
+        .. literalinclude:: ../examples/matrix_mult.py
```
Comment (jmfranck) [inline pyspecdata/core.py:3562]: if you build the sphinx, does this show up correctly?

File: pyspecdata/core.py [hunk 9]
```diff
@@ -3538,32 +3558,67 @@ def dot(self,arg):
...
+        logger.debug(strm("add a new axis",A.data.shape))
+        logger.debug(strm("add a new axis",B.data.shape))
+        A.data = moveaxis(A.data,ax_idx,-1) # columns position
+        B.data = moveaxis(B.data,ax_idx,-2) # row position
+        logger.debug(strm("after movement",A.data.shape))
+        logger.debug(strm("after movement",B.data.shape))
+        self.data = matmul(A.data,B.data)
+        logger.debug(strm("after mult",self.data.shape))
+        self.data = self.data[...,0,0]
+        logger.debug(strm("remove extras",self.data.shape))
```
Comment (jmfranck) [inline pyspecdata/core.py:3614]: @bett8500 : so did we establish that matmul and dot were the same speed or different speeds?  What about for 3D and higher vs 2D?

File: pyspecdata/core.py [hunk 10]
```diff
@@ -4565,99 +4635,100 @@ def nnls(self, dimname, newaxis_dict, kernel_func, l=0):
...
+        S1 = S1[0:s1]
+        V1 = V1[0:s1,:]
+        S1 = S1*eye(s1)
+        logger.debug(strm('Compressed SVD of K1:',[x.shape for x in (U1,S1,V1)]))
+        #{{{ prepping 2D
+        if twoD:
+            U2 = (svd_return[1])[0]
+            S2 = (svd_return[1])[1]
+            V2 = (svd_return[1])[2]
             s2 = where(S2 > default_cut)[0][-1]
```
Comment (jmfranck) [inline pyspecdata/core.py:4717]: @aabeaton : just checking that this branch is the most recent version of the code.  I thought maybe we landed on a list of U matrices, list of V matrices, etc. (list can have 1 or 2 elements) -- or is this just done for the kernel, or not at all, or is this not eh most recent version of the code?

File: pyspecdata/core.py (hunk 6 above)
Comment (jmfranck) [reply pyspecdata/core.py:2600]: Tried it in ipython, and I'm wrong about that
```
In [14]: py_seed(1432)

In [15]: print(a.add_noise(0.5))
[1.07955776, 0.07674935, 1.3582009 , 2.53816454, 4.53411609
 5.04623931, 4.89612688, 6.06584429, 7.21472772, 8.39430309]
                +/-None
        dimlabels=['t']
        axes={`t':[0., 1., 2., 3., 4., 5., 6., 7., 8., 9.]
                        +/-None}


In [16]: py_seed(1432)

In [17]: print(a.add_noise(0.5))
[1.02877369, 0.05866427, 1.18729449, 2.40708142, 4.8559464
 5.16286492, 4.50150268, 5.66294507, 6.66677689, 8.21547053]
                +/-None
        dimlabels=['t']
        axes={`t':[0., 1., 2., 3., 4., 5., 6., 7., 8., 9.]
                        +/-None}
```
(I would expect both to be the same number if seeded in the same way)

Comment (jmfranck) [reply pyspecdata/core.py:2600]: What I'm saying here is correct ... notice below how seeding the same number results in the same random sequence following the seed, even though two subsequent calls (without a seed between) give back different values:
```
In [18]: a = nddata(r_[0:10.],'t')

In [19]: py_seed(1432)

In [20]: print(a.C.add_noise(2))
[-0.20313629,  0.92765967,  1.31637436,  2.47566753,  5.28732126
  5.46650244,  4.4215032 ,  5.38840311,  5.80819667,  8.28466978]
                +/-None
        dimlabels=['t']
        axes={`t':[0., 1., 2., 3., 4., 5., 6., 7., 8., 9.]
                        +/-None}


In [21]: print(a.C.add_noise(2))
[-0.19759952,  1.7238222 ,  0.60702088,  2.93038397,  3.71326082
  3.65133217,  6.60735107,  7.37541355,  7.94590838,  6.72645224]
                +/-None
        dimlabels=['t']
        axes={`t':[0., 1., 2., 3., 4., 5., 6., 7., 8., 9.]
                        +/-None}


In [22]: py_seed(1432)

In [23]: print(a.C.add_noise(2))
[-0.20313629,  0.92765967,  1.31637436,  2.47566753,  5.28732126
  5.46650244,  4.4215032 ,  5.38840311,  5.80819667,  8.28466978]
                +/-None
        dimlabels=['t']
        axes={`t':[0., 1., 2., 3., 4., 5., 6., 7., 8., 9.]
                        +/-None}

```

File: pyspecdata/core.py [hunk 11]
```diff
@@ -5116,7 +5138,7 @@ def multimin(self,minfunc,axisname,filterwidth,numberofmins):
             #}}}
     def repwlabels(self,axis):
         return None
-    def add_noise(self,intensity):
+    def add_noise(self,intensity,seed=None):
         '''Add Gaussian (box-muller) noise to the data.
```
Comment (jmfranck) [inline pyspecdata/core.py:5142]: so I would get rid of this

File: pyspecdata/core.py [hunk 12]
```diff
@@ -5130,7 +5152,7 @@ def add_noise(self,intensity):
         if isinstance(intensity, type(emptyfunction)):
             intensity = intensity(lambda x: self.data)
         return_complex = iscomplexobj(self.data)
-        self.data += box_muller(self.data.size, return_complex=return_complex).reshape(self.data.shape) * intensity
+        self.data += box_muller(self.data.size, return_complex=return_complex, seed=seed).reshape(self.data.shape) * intensity
         return self
```
Comment (jmfranck) [inline pyspecdata/core.py:5156]: and get rid of this

File: demos/nddata_NNLS_Tikhonov_190104.py (hunk 3 above)
Comment (aabeaton) [reply demos/nddata_NNLS_Tikhonov_190104.py:282]: I do not know exactly what this is or why it was commented out. But this code as it stands runs on my computer -- it takes some time because it is doing the timing thing that you had put in at some point a while back.

Comment (aabeaton) [reply demos/nddata_NNLS_Tikhonov_190104.py:282]: I know @guinneak198 got stuff to work wrt demos/sphinx(?) so maybe if she and I have a brief meeting I can try to implement all that on what is here and the other items in the demos folder... there is also an examples folder, which makes things a little confusing

File: examples/BRD_test.py (hunk 4 above)
Comment (aabeaton) [reply examples/BRD_test.py:12]: I believe the script in "demos" was meant to just test the threading capabilities when that had been implemented, also to assess whether it was saving us time or not

File: pyspecdata/core.py (hunk 6 above)
Comment (aabeaton) [reply pyspecdata/core.py:2632]: Ok, that is an interesting example. So to clarify, you're saying what was in place before this is all that is needed, and what I have added with the seed=None and the if statement can all be deleted?

File: demos/nddata_NNLS_Tikhonov_190104.py (hunk 3 above)
Comment (jmfranck) [reply demos/nddata_NNLS_Tikhonov_190104.py:282]: that's a good idea, if it's a quick thing.  There should be no separate demos folder, I think, just an examples folder.  None of the demos should be intended for ipynb, only stand-alone scripts

File: examples/BRD_test.py (hunk 4 above)
Comment (jmfranck) [reply examples/BRD_test.py:12]: ahh -- that makes sense -- if we wanted to make an example to test the threading capabilities, we could do that.

File: pyspecdata/core.py (hunk 6 above)
Comment (jmfranck) [reply pyspecdata/core.py:2632]: I think so.  I'm saying there's no need to reference the seed inside the individual functions -- you can do that just once in th main script.

File: demos/nddata_NNLS_Tikhonov_190104.py (hunk 3 above)
Comment (guinneak198) [reply demos/nddata_NNLS_Tikhonov_190104.py:282]: At this point I think we are more concerned with getting this PR merged for processing the ILT's of D2O data. I think examples could be a separate pull request/wrike task to organize things a bit better?

File: examples/matrix_mult.py [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    return s
+s = find_file('15N_S175R1a_pR_new_9G_200309.DSC',
+        exp_type='francklab_esr/Sam',
+        postproc=preproc_ESR_multiharmonic_centered)
+man_calc = []
+for j in range(ndshape(s)['harmonic']):
+    man_calc.append((s['harmonic',j].data**2).sum())
+result = s.along('$B_0$') @ s
+print("find the norm squared", result)
+assert all(isclose(array(man_calc), result.data))
```
Comment (guinneak198) [inline examples/matrix_mult.py:37]: This doesn't pertain to this pull request and was Sam's project, not sure how we can/should use it here

File: pyspecdata/core.py [hunk 14]
```diff
@@ -3630,14 +3633,19 @@ def along(self,dimname):
         multiplication (represents the rows/columns)."""
         self._matmul_along = dimname
         return self
-    def along(self,dimname):
-        """Specifies the dimension for the next matrix
-        multiplication (represents the rows/columns)."""
-        self._matmul_along = dimname
-        return self
-    #@profile
```
Comment (guinneak198) [inline pyspecdata/core.py:3638]: matrix stuff (sam?) not needed for this PR

File: pyspecdata/core.py (hunk 8 above)
Comment (guinneak198) [reply pyspecdata/core.py:3664]: not sure if this pertains to nnls or tikhonov

File: demos/nddata_NNLS_Tikhonov_190104.py (hunk 3 above)
Comment (guinneak198) [reply demos/nddata_NNLS_Tikhonov_190104.py:282]: This example takes forever to run. I'm adding it to examples and removing it from here but we should consider the time consumption

File: examples/BRD_test.py [hunk 15]
```diff
@@ -0,0 +1,103 @@
+# for 1D BRD, adapted mainly from Venkataramanan 2002
```
Comment (guinneak198) [inline examples/BRD_test.py:1]: This example works

File: pyspecdata/core.py [hunk 16]
```diff
@@ -3630,14 +3631,19 @@ def along(self,dimname):
...
+        it will dot along that dimension
+        (take a matrix multiplication where that dimension represents the columns of ``self`` and the rows of ``arg``)
+
+        Note that if you have your dimensions named "rows" and "columns", this
+        will be very confusing, but if you have your dimensions named in terms
+        of the vector basis they are defined/live in, this makes sense.
+
+        If there are zero or no matching dimensions, then use
+        :func:`~pyspecdata.nddata.along` to specify the dimensions for matrix
+        multiplication / dot product.
```
Comment (guinneak198) [inline pyspecdata/core.py:3646]: all this is required for the example BRD_test.py

File: pyspecdata/core.py [hunk 17]
```diff
@@ -3630,14 +3631,19 @@ def along(self,dimname):
         multiplication (represents the rows/columns)."""
         self._matmul_along = dimname
         return self
-    def along(self,dimname):
-        """Specifies the dimension for the next matrix
-        multiplication (represents the rows/columns)."""
-        self._matmul_along = dimname
-        return self
-    #@profile
```
Comment (jmfranck) [inline pyspecdata/core.py:3638]: why is this deleted?

Comment (jmfranck) [inline pyspecdata/core.py:3638]: why is this deleted?

File: pyspecdata/core.py [hunk 18]
```diff
@@ -3652,32 +3658,67 @@ def dot(self,arg):
...
+        if hasattr(self,'_matmul_along'):
+            del(self._matmul_along)
+        if hasattr(arg,'_matmul_along'):
+            del(arg._matmul_along)
+        # }}}
+        if dot_dim_B != dot_dim_A:
+            arg = arg.C.rename(dot_dim_B,dot_dim_A)
         A,B = self.aligndata(arg)
-        matching_dims = list(set(self.dimlabels) & set(arg.dimlabels))
-        assert len(matching_dims) > 0, "no matching dimensions!"
```
Comment (jmfranck) [inline pyspecdata/core.py:3658]: why is this removed?

**Review by jmfranck — CHANGES_REQUESTED** [review]
In addition to the comments here, I want to make sure that this includes a working 2D ILT example (i.e. where both dimensions are relaxometry dimensions -- *not* chemical shift).
